### PR TITLE
[fix] Make scale dependent visibility more robust when handling non-round scales (denominators)

### LINKED ISSUE #58150: Scale Dependent Visibility on layer wrong on scale interval edges in EPSG:25832
### What is the bug or the crash?

I have two layers, L1 and L2, and both has `Scale Dependent Visibility` (Layer Properties / Rendering) enabled:
- L1 has `Maximum (inclusive)` set to `1:1` and has `Minimum (exclusive)` set to `1:100`.
- L2 has `Maximum (inclusive)` set to `1:100` and has `Minimum (exclusive)` set to `1:1,000`.

When setting the zoom level of QGIS to exactly `1:100` both L1 and L2 are visible. I would expect L1 to be invisible (since it has exclusive on Minimum) and L2 to be visible (since it has inclusive on Maximum).

I didn't submit this ticket just because of issues with scale `1:100`. I see the same issue at scale `1:10,000`, `1:25,000`, `1:100,000`, `1:250,000` and `1:1,000,000` but not at scale `1:50,000`, `1:500,000` and `1:5,000,000`.
### Steps to reproduce the issue

Add two layers and set the scale dependent visibility of both layers according to the bug-description.

### Versions

3.34.8, 3.38.0, 3.39 (67e056379e) - all on Windows 10.

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

Looking briefly at the code I see this:

https://github.com/qgis/QGIS/blob/9e47ca4230edfe5d72bc2130d5d584086f3a8402/src/core/qgis.cpp#L56
https://github.com/qgis/QGIS/blob/9e47ca4230edfe5d72bc2130d5d584086f3a8402/src/core/qgsmaplayer.cpp#L1147-L1151

Could the `SCALE_PRECISION` be the reason that both layers are rendered at exactly `1:100` in EPSG:25832?

### PR DESCRIPTION
**Pull request by gacarrillor**
Follow-up https://github.com/qgis/QGIS/pull/58568/commits/b8f976012ab383e29e7156ff75ea3ea836b99e06

Since the map scale (denominator) is a double, it could be a non-round number, which needs to be considered when comparing scales for scale dependent visibility of features, labels, diagrams, and the like. This should be done for both range limits: minimum (exclusive) and maximum (inclusive) scales.

This PR:

 + Harmonizes scaled dependent visibility for map layer rendering, rule-based rendering, rule-based labeling, single and stacked diagrams and app's 'Zoom to Visible Scale' action.
 + Makes sure the **inclusive** range limit (maximum scale) is taken into account.
 + Suggests to use `qgsDoubleNear()` for equal checks, thus taking non-round scales into account.
 + Deprecates the `Qgis::SCALE_PRECISION` workaround.
 + Adds more tests for `QgsMapLayer->isInRange()`.
 + Adds 2 handy methods `equalToOrGreaterThanMinimumScale()` and `lessThanMaximumScale()` to make scale conditions more readable, with corresponding tests.

Sample of inconsistent behavior fixed by this PR:

### Map layer

Expected: Layer shouldn't be rendered if map scale is a non-round number _equals_ to layer's `minimumScale` (exclusive).
Current (screenshot): Range: 25000-50000, map scale: 49999,999999999716. The layer is rendered.
Note the map scale combobox shows 50000 as the current map scale (denominator). 

Screenshot from 2024-09-26 18-19-20_maplayer_min_non_exclusive_non_round

### Rule-based labeling

Expected: Labels within a rule should be rendered if map scale is a non-round number _equals_ to rule's `maximumScale`(inclusive).
Current (screenshot): Rule's scale range: 50000-100000, map scale: 49999.99999999976. The labels are not rendered.
Note the map scale combobox shows 50000 as the current map scale (denominator). 

Screenshot from 2024-09-30 20-59-00_rule_based_labeling_max_non_inclusive_non_round

Similar issues can be reproduced for rule-based rendering.

Note: Those non-round map scales can be consistently obtained by resizing the QGIS main window (Ubuntu Linux) and resetting the scale to a round number via main Scale combobox. Users think the scale is still round (according to the combobox), but internally, it's a non-round number (in practice, equal to the round one).

Making scale depentent visibility more robust when handling non-round numbers would also cover use cases where, e.g., plugin devs set non-round scales via PyQGIS, as a result of a calculation.

Additional fixes:
Fix #58150

Other users have reported this longstanding issue, for instance:
https://github.com/qgis/QGIS/pull/2863#issuecomment-194144418
https://github.com/qgis/QGIS/issues/42443

### CODE REVIEW

File: src/core/qgis.h [hunk 1]
```diff
@@ -6256,6 +6258,34 @@ CORE_EXPORT bool qgsVariantEqual( const QVariant &lhs, const QVariant &rhs );
...
+ *
+ * \param scale The current scale to be compared.
+ * \param minScale The minimum map scale (i.e. most "zoomed out" scale) at
+ * which features, labels or diagrams will be visible. The scale value
+ * indicates the scale denominator, e.g. 1000.0 for a 1:1000 map.
+ * \see qgsLessThanMaximumScale()
+ *
+ * \since QGIS 3.40
+ */
+CORE_EXPORT bool qgsEqualToOrGreaterThanMinimumScale( const double scale, const double minScale );
```
Comment (nyalldawson) [inline src/core/qgis.h:6273]: Could this be moved to static function in a "QgsScaleUtils" class instead?

Comment (gacarrillor) [reply src/core/qgis.h]: Yes, done in https://github.com/qgis/QGIS/pull/58968/commits/2b4debadb0f08507411bb5c147c1c6348ac51aea

**Comment by gacarrillor** [comment]
@nyalldawson, should I also implement this for the label's `scale dependent visibility` widget?

I didn't yet because there was no visual hint (e.g., in the tool tip) on whether the maximum scale value should be inclusive.

image

**Review by nyalldawson — APPROVED** [review]
Nice cleanups, thank you!

**Comment by nyalldawson** [comment]
@gacarrillor 
> should I also implement this for the label's scale dependent visibility widget?

Yes, let's make it consistent throughout! Feel free to open that in a new PR.

**Pull request merged**